tools/topology/topology1/sof-icl-rt711-rt1308-rt715-hdmi.m4

```diff
@@ -31,6 +31,18 @@ ifdef(`AMP_2_LINK',`',
 ifdef(`MIC_LINK',`',
 `define(MIC_LINK, `3')')
 
+# UAJ ID: 0, 1
+# AMP ID: 2, 3 (if EXT_AMP_REF defined)
+# DMIC ID: 4
+# HDMI ID calculated based on the configuraiton
+ifdef(`NO_LOCAL_MIC',
+  `ifdef(`NOAMP',
+       `ifdef(`NOJACK', `define(HDMI_BE_ID_BASE, `0')', `define(HDMI_BE_ID_BASE, `2')')',
+       `ifdef(`EXT_AMP_REF', `define(HDMI_BE_ID_BASE, `4')', `define(HDMI_BE_ID_BASE, `3')')'
+   )',
+   `define(HDMI_BE_ID_BASE, `5')'
```

Comment:
do we really have to do this three-level tests, or could we use the same technique as in the other topology file:

````
ifdef(`NO_JACK',
`
define(JACK_OFFSET, `0')
',
`
define(JACK_OFFSET, `2')
')


# if there is an external RT1308 amplifier connected over SoundWire,
# enable "EXT_AMP" option in the CMakefile.
ifdef(`EXT_AMP',
`
define(AMP_OFFSET, `1')
',
`
define(AMP_OFFSET, `0')
'
)

ifdef(`EXT_AMP_REF',
`
define(AMP_REF_OFFSET, `1')
',
`
define(AMP_REF_OFFSET, `0')
'
)
define(HDMI_BE_ID_BASE, eval(JACK_OFFSET+AMP_OFFSET+AMP_REF_OFFSET+DMIC_OFFSET))
````

Reply:
sof-tgl-rt711-rt1308 is something wrong. But it currently works because it doesn't meet the corner case. Let's image we don't have Jack and DMIC but have smart AMP on the board. Our driver code will always assigned the AMP BE ID from 2. So in this configuration, the driver will get the HDMI ID from 4, but in the sof-tgl-rt711-rt1308 tplg, HDMI ID will be from 2.

Reply:
@plbossart Currently, we don't reserve a dai link for amp reference. But we do have a fixed dsi link id for sdw mic.
So the hdmi link will always be 5 if there is a sdw mic present, but it will be 4 if we use a amp without reference.

@libinyang How about
```
define(HDMI_BE_ID_BASE, `0')

ifdef(`NO_JACK', `'
	` define(HDMI_BE_ID_BASE, `2') '
)

ifdef(`NOAMP', `'
	` define(HDMI_BE_ID_BASE, `3') '
)

ifdef(`EXT_AMP_REF',
	` define(HDMI_BE_ID_BASE, `4') ',
	`'
)

ifdef(`NO_LOCAL_MIC', `'
	` define(HDMI_BE_ID_BASE, `5') '
)
```
We set HDMI_BE_ID_BASE = 0 initially, and check headset, amp, reference, mic one by one and overwrite HDMI_BE_ID_BASE when we find any sdw link.

Reply:
@plbossart The PR test has been passed.

Reply:
@libinyang you didn't reply to @bardliao 's suggestion

Reply:
> @libinyang you didn't reply to @bardliao 's suggestion

@plbossart @bardliao Oh, sorry. I thought I have replied @bardliao 's suggestion. But it seems I failed to send out my reply. The suggestion can't handle: 'NOAMP' is set but 'NO_LOCAL_MIC' is not set. In this case, there is local MIC, so it HDMI BE ID should be 5, but Bard's suggestion will get 3 for HDMI BE ID.

Reply:
> > @libinyang you didn't reply to @bardliao 's suggestion
> 
> @plbossart @bardliao Oh, sorry. I thought I had replied @bardliao 's suggestion. But it seems I failed to send out my reply. The suggestion can't handle: 'NOAMP' is set but 'NO_LOCAL_MIC' is not set. In this case, there is local MIC, so HDMI BE ID should be 5, but Bard's suggestion will get 3 for HDMI BE ID.

@libinyang Won't the HDMI BE ID be 5 in this case? HDMI_BE_ID_BASE will be overwrite to 5 if NO_LOCAL_MIC is not set, no?

Reply:
@bardliao Oh, my fault. There's a little fault in your suggestion. We have to undefine  HDMI_BE_ID_BASE before redefining it. Please see my new PR.